### PR TITLE
[wallet] Close DB on error.

### PR DESCRIPTION
**Pull request by kallewoof**
This PR intends to plug some leaks. It specifically implements adherence to the requirement in BDB to close a handle which failed to open (https://docs.oracle.com/cd/E17276_01/html/api_reference/C/dbopen.html):

> The `DB->open()` method returns a non-zero error value on failure and 0 on success. If `DB->open()` fails, the `DB->close()` method must be called to discard the DB handle.

### CODE REVIEW

**Comment by practicalswift** [comment]
Concept ACK! Nice find 👍

**Review by gmaxwell — APPROVED** [review]
utACK

**Comment by MarcoFalke** [comment]
utACK 03bc719a85cb4928cb4b43d0bc4142f72cb01b23

**Comment by paveljanik** [comment]
utACK 03bc719

File: src/wallet/db.cpp [hunk 1]
```diff
@@ -536,8 +538,10 @@ bool CDB::Rewrite(CWalletDBWrapper& dbw, const char* pszSkip)
                         env->CloseDb(strFile);
                         if (pdbCopy->close(0))
                             fSuccess = false;
-                        delete pdbCopy;
+                    } else {
+                        pdbCopy->close(0);
                     }
+                    delete pdbCopy;
```
Comment (laanwj) [inline src/wallet/db.cpp]: Shouldn't it always close pdbCopy before deleting it, not just on failure?

Comment (kallewoof) [inline src/wallet/db.cpp]: Line #537 closes for fSuccess being false and #542 for true. Am I missing a case?

**Review by promag — COMMENTED** [review]
utACK 03bc719.

BTW, how did you catch this?

**Comment by kallewoof** [comment]
@promag I ran bitcoind through Instruments on a mac, and noticed I think 40k leaks (not 40 kb, 40 thousand occurrences..) on startup. One of the traces for leaked memory was the DB open call.

**Comment by jonasschnelli** [comment]
utACK 03bc719a85cb4928cb4b43d0bc4142f72cb01b23

**Pull request merged**